### PR TITLE
dexc-desktop: set the window icon and make Windows build work

### PR DESCRIPTION
**Pull request by chappjc**
image => image

### CODE REVIEW

**Comment by chappjc** [comment]
Two other fixes:
- Windows must is ico, not png.  Do like with dexc systray.
- Don't use a webview process to open the logs folder.  Use the system shell execution.

**Comment by chappjc** [comment]
dexc-desktop is now working on Windows.

Added some build notes in build.cmd (https://github.com/decred/dcrdex/pull/2351/commits/3aed08cf118cc51c86f633c297fe47afa120cef1), but we might want to make more of a script at some point.
You may still need to install the [WebView2 runtime](https://developer.microsoft.com/en-us/microsoft-edge/webview2/) unless you're on Windows 11.
For GCC on windows, I've tested with the ucrt64 mingw64 from https://www.msys2.org/.  Install as the docs describe, using `pacman -S base-devel mingw-w64-ucrt-x86_64-gcc` to get the urct64 compilers.  Do a full update with `pacman -Suy`.  Then edit your Window's *user's* `PATH`, appending something like `;C:\msys64\ucrt64\bin` to the string.  Then in a regular windows terminal, test `gcc --version`.

File: client/cmd/dexc-desktop/icon.go [hunk 1]
```diff
@@ -6,7 +6,14 @@
...
 	"github.com/webview/webview"
 )
 
+//go:embed imgs/dexc.png
+var FavIcon []byte
+
+//go:embed imgs/symbol-bw-round.png
+var SymbolBWIcon []byte
+
 func useIcon(w webview.WebView, iconPath string) { /* not supported on this platform */ }
```
Comment (chappjc) [inline client/cmd/dexc-desktop/icon.go:19]: Oh, I meant to write the embedded `FavIcon` to a temp file path for this.

**Comment by JoeGruffins** [comment]
Attempting to build with the script:
```
C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop>go build -v -ldflags="-H windowsgui"
github.com/webview/webview
# github.com/webview/webview
In file included from webview.h:1056,
                 from webview.cc:1:
C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop\libs\webview2\build\native\include/WebView2.h:1399:10: fatal error: EventToken.h: No such file or directory
 1399 | #include "EventToken.h"
      |          ^~~~~~~~~~~~~~
compilation terminated.
```

Trying to download webview 2 from your link tells me I already have it. Should I troubleshoot this or is it obvious whats wrong? Im on Windows 10 Home

**Comment by JoeGruffins** [comment]
Googling, I thought maybe my windows SDK was old? But I am using 10.0.22000 which is newer I think.

**Comment by chappjc** [comment]
@JoeGruffins try a different gcc.  That's a pretty old version of GCC, plus tdm isn't tested with webview (https://github.com/webview/webview#mingw-w64-requirements)
I used msys2's ucrt64 version as described here: https://github.com/decred/dcrdex/pull/2351#issuecomment-1549963442
But I also tested that with windows 11 and I didn't need any windows sdk, only the webview sdk downloaded with the nuget comment.

**Review by JoeGruffins — APPROVED** [review]
Ah working now! Everything looks ok on Windows 10. I think this may have been my problem, I didn't notice before:

```
curl -sSL "https://www.nuget.org/api/v2/packag
e/Microsoft.Web.WebView2"   | tar -xf - -C libs\webview2
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
```

I was able to open the archive using `unzip` however.

File: client/cmd/dexc-desktop/gen.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+//go:generate go run github.com/tc-hib/go-winres@v0.3.1 make --in winres.json --arch "386,amd64"
+
+package main
+
+// After generating the rsrc_windows_amd64.syso file, it will be included in the
+// binary when making an amd64 build for windows.
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/cmd/dexc-desktop/gen.go:9]: `go build` picks up syso files and includes them in the binary.  In this case it's a way to get windows resources into the executable.  This syso file is a special windows resource file.  This is the stuff the Windows ABI understands, giving it e.g. the info in the Version table when you right click a file and see properties in Windows Explorer, give a list of icons the explorer can read and use, etc.

The naming structure `blah_<os>_<arch>` is recognize by `go build` for conditional inclusion of files.  You can do the same thing for .go files instead of simple go tags.

File: client/cmd/dexc-desktop/main.go [hunk 3]
```diff
@@ -529,8 +533,9 @@ func systrayOnReady(ctx context.Context, logDirectory string, openC chan<- struc
 		mLogs := systray.AddMenuItem("Open logs folder", "Open the folder with your DEX logs.")
 		go func() {
 			for range mLogs.ClickedCh {
-				log.Debug("Opening browswer to log directory at", logDirURL)
-				runWebviewSubprocess(ctx, logDirURL)
+				if err := browser.OpenURL(logDirURL); err != nil {
+					fmt.Fprintln(os.Stderr, err)
```
[the remark this answers is not in the document]
Comment (chappjc) [reply client/cmd/dexc-desktop/main.go:537]: The context here is you're trying to open the logs folder but it's not able to.  (In this situation you're actually looking for the log file, but it's failing for some reason).  I'll print and log.

File: client/cmd/dexc-desktop/gen.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply client/cmd/dexc-desktop/gen.go:9]: For reference about the file naming structure:  https://pkg.go.dev/cmd/go#hdr-Build_constraints

> If a file's name, after stripping the extension and a possible _test suffix, matches any of the following patterns:
> 
> *_GOOS
> *_GOARCH
> *_GOOS_GOARCH
> (example: source_windows_amd64.go) where GOOS and GOARCH represent any known operating system and architecture values respectively, then the file is considered to have an implicit build constraint requiring those terms (in addition to any explicit constraints in the file).

Since this isn't a .go file where we can use build tags, we use the file name instead.

**Comment by chappjc** [comment]
> Ah working now! Everything looks ok on Windows 10. I think this may have been my problem, I didn't notice before:
> 
> ```
> curl -sSL "https://www.nuget.org/api/v2/packag
> e/Microsoft.Web.WebView2"   | tar -xf - -C libs\webview2
> tar: This does not look like a tar archive
> tar: Skipping to next header
> tar: Exiting with failure status due to previous errors
> ```
> 
> I was able to open the archive using `unzip` however.

Maybe just need to update windows.  Supposedly tar in Windows landed in 2018.  https://www.addictivetips.com/windows-tips/use-tar-on-windows-10/

**Comment by JoeGruffins** [comment]
> Maybe just need to update windows. 

To Windows 11? I'm sure I'm on whatever the latest Windows 10 is. Also tar is installed there, but said the archive was the wrong type. May this was no window's tar but the msys2 version? Anyway it worked in the end.

**Comment by chappjc** [comment]
> Maybe just need to update windows. To Windows 11? I'm sure I'm on whatever the latest Windows 10 is.

Nah, I just meant Windows Update to the latest revision of Windows 10.

> Also tar is installed there, but said the archive was the wrong type. May this was no window's tar but the msys2 version? Anyway it worked in the end.

Yeah, nbd.

**Comment by chappjc** [comment]
Going to assume it's just Joe and I testing the windows build.sh script for dexc-desktop.

**Comment by chappjc** [comment]
Added instructions in build.cmd. https://github.com/decred/dcrdex/blob/1a173d801c495ce7127d12c00dffcba1432dfdbc/client/cmd/dexc-desktop/build.cmd#L1
This has worked in three separate windows installs, two windows 10 and the other windows 11.

Also rebased after the macdriver merge.  A little tricky.  Need to retest.

File: client/cmd/dexc-desktop/icon_darwin.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+//go:build darwin
```
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply client/cmd/dexc-desktop/icon_darwin.go:4]: yh.

**Comment by chappjc** [comment]
Retested on Windows.  Also verified that the dexc-desktop build still works on Linux and Darwin, although this PR does not apply to Darwin.

**Review by buck54321 — COMMENTED** [review]
Getting closer. Started with fresh Windows install and followed directions. Build went fine, but wouldn't run. From logs:
```
2023-07-21 22:37:46.750 [INF] APP: Opening new window. 1 windows open now
2023-07-21 22:37:55.476 [ERR] APP: webview error: exit status 2
2023-07-21 22:37:55.476 [INF] APP: Closing window. 0 windows remain open.
```

**Comment by chappjc** [comment]
Qs:
- Can you try with and without the WebView2Loader.dll in the same folder as dexc-desktop.exe?
- What windows ver and revision? I'm wondering if you need to install the runtime: https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section

File: client/cmd/dexc-desktop/build.cmd [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+::      from that folder with `npm ci` and `npm run build`. See the Node.js
+::      install instructions above.
+::
+:: NOTE: On every machine tested, the Webview2 *runtime* was already installed,
+:: but if you have removed it or Edge itself, it may be necessary to install:
+:: https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section 
+
+echo "Fetching and extracting the latest WebView2 SDK..."
+mkdir libs\webview2
+curl -sSL "https://www.nuget.org/api/v2/package/Microsoft.Web.WebView2" | tar -xf - -C libs\webview2
```
Comment (chappjc) [inline client/cmd/dexc-desktop/build.cmd:52]: I've been wondering if it's necessary to pin to a specific version. e.g. https://www.nuget.org/packages/Microsoft.Web.WebView2/1.0.1823.32

**Comment by buck54321** [comment]
> Qs:
> 
> * Can you try with and without the WebView2Loader.dll in the same folder as dexc-desktop.exe?
> * What windows ver and revision? I'm wondering if you need to install the runtime: https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section

Moving the WebView2Loader.dll file didn't work. Trying to install the webview2 runtime from provided link, but download button is busted or something. Grrr.

**Comment by buck54321** [comment]
Installing the runtime didn't seem to help.

**Comment by buck54321** [comment]
Wait. I tried again and it opened. Weird.

**Comment by chappjc** [comment]
Hurray!  I hope it's not always that flakey.

**Comment by buck54321** [comment]
Fresh Windows install again and hit some issues. I don't think the WebView runtime system installation step is a prerequisite for building. Just for running on Windows 10. In fact, after running through the newly modified instructions, I was getting `webview error: exit status 2` again until I deleted (well, renamed) the WebView2Loader.dll file, which seems like a clue. After deleting that library, everything seemed to be running and I was able to open multiple dexc windows, but there was nothing rendering. Just a blank white screen. Ugh.

**Comment by ukane-philemon** [comment]
> Fresh Windows install again and hit some issues. I don't think the WebView runtime system installation step is a prerequisite for building. Just for running on Windows 10. In fact, after running through the newly modified instructions, I was getting `webview error: exit status 2` again until I deleted (well, renamed) the WebView2Loader.dll file, which seems like a clue. After deleting that library, everything seemed to be running and I was able to open multiple dexc windows, but there was nothing rendering. Just a blank white screen. Ugh.

I did everything in the build guide but after creating my dexc password on a fresh Win 10 install, it refused to display other pages. 
Found this in the console: Maybe some js methods are not supported on Windows webview? I'm also going to test on Win 11.

Screenshot 2023-11-08 at 3 44 34 PM

**Comment by buck54321** [comment]
Those errors don't mean that the JS functions don't exist, they mean we are passing undefined elements. Are you rebased on master?

**Comment by ukane-philemon** [comment]
Oh, I had to build on an updated branch then cherry-picked from the Windows VM. npm kept failing to run a successful ci.

**Comment by ukane-philemon** [comment]
I'll try again tomorrow.

**Review by ukane-philemon — COMMENTED** [review]
Looks good. Tested on Win 10 and 11. Dexc-desktop had no display issues.

A non-blocking issue I noticed:

Some Windows devices (a small percentage according to Microsoft) do not have the Webview2 Runtime but the good news is that we can handle this edge case: https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution#details-about-the-webview2-runtime

Or state in the release notes or somewhere that only updated Win 10 and Win 11 are supported(these have the Webview2 runtime installed).

The Webview package installed and embedded by the build script is required (got the error below on both Win 10 and 11 when I tried to build without embedding the Webview2 package):
Screenshot 2023-11-10 at 5 09 30 PM

In conclusion, we can require the Webview2 runtime (https://learn.microsoft.com/en-us/microsoft-edge/webview2/samples/deployment-samples) or detect this during installation (https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/distribution#detect-if-a-suitable-webview2-runtime-is-already-installed) if we use a native Windows build&install process.

PS: If you’re using a Windows VM to test, you should turn off the Auto Detect proxy settings, else npm will always mess things up.

Screenshot 2023-11-10 at 4 59 05 PM

**Comment by buck54321** [comment]
Superceded by #2635
